pkg/operator/workloads/logs.go

```diff
@@ -40,423 +37,187 @@ const (
-	ContainerName string
-	RestartCount  int32
-}
-
-func (l LogKey) String() string {
-	return fmt.Sprintf("%s,%s,%d", l.PodName, l.ContainerName, l.RestartCount)
-}
-
-func StringToLogKey(str string) LogKey {
-	split := strings.Split(str, ",")
-	restartCount, _ := s.ParseInt32(split[2])
-	return LogKey{PodName: split[0], ContainerName: split[1], RestartCount: restartCount}
-}
-
-func GetLogKey(pod kcore.Pod, status kcore.ContainerStatus) LogKey {
-	return LogKey{PodName: pod.Name, ContainerName: status.Name, RestartCount: status.RestartCount}
-}
-
-func CurrentLoggingProcesses(logProcesses map[string]*os.Process) strset.Set {
-	set := strset.New()
-	for identifier := range logProcesses {
-		set.Add(identifier)
-	}
-	return set
-}
-
-func GetLogKeys(pod kcore.Pod) strset.Set {
-	containerStatuses := append(pod.Status.InitContainerStatuses, pod.Status.ContainerStatuses...)
-	logKeys := strset.New()
-	for _, status := range containerStatuses {
-		if status.State.Terminated != nil || status.State.Running != nil || (status.State.Waiting != nil && status.RestartCount != 0) {
-			logKeys.Add(GetLogKey(pod, status).String())
+func pumpStdin(socket *websocket.Conn) {
+	socket.SetReadLimit(socketMaxMessageSize)
+	for {
+		_, _, err := socket.ReadMessage()
+		if err != nil {
+			break
 		}
 	}
-
-	return logKeys
 }
 
-func createKubectlProcess(logKey LogKey, attrs *os.ProcAttr) (*os.Process, error) {
-	cmdPath := "/usr/local/bin/kubectl"
-
-	kubectlArgs := []string{"kubectl", "-n=" + config.Cortex.Namespace, "logs", "--follow=true", logKey.PodName, logKey.ContainerName, fmt.Sprintf("--tail=%d", initLogTailLines)}
-	process, err := os.StartProcess(cmdPath, kubectlArgs, attrs)
-	if err != nil {
-		return nil, errors.Wrap(err, strings.Join(kubectlArgs, " "))
-	}
-	return process, nil
+type FluentdLog struct {
+	Log string `json:"log"`
 }
 
-func podCheck(podCheckCancel chan struct{}, socket *websocket.Conn, initialPodList []kcore.Pod, wrotePending bool, inr *os.File) {
+func StreamFromCloudWatch(podCheckCancel chan struct{}, appName string, podLabels map[string]string, socket *websocket.Conn) {
 	timer := time.NewTimer(0)
 	defer timer.Stop()
 
-	processMap := make(map[string]*os.Process)
-	defer deleteProcesses(processMap)
-	labels := initialPodList[0].GetLabels()
-	podSearchLabels := map[string]string{
-		"appName":      labels["appName"],
-		"workloadType": labels["workloadType"],
-		"userFacing":   "true",
-	}
-
-	if labels["workloadType"] == workloadTypeAPI {
-		podSearchLabels["apiName"] = labels["apiName"]
-	} else {
-		podSearchLabels["workloadID"] = labels["workloadID"]
-	}
-
-	outr, outw, err := os.Pipe()
-	if err != nil {
-		errors.Panic(err, "logs", "kubectl", "os.pipe")
-	}
-	defer outw.Close()
-	defer outr.Close()
-
-	procAttr := os.ProcAttr{
-		Files: []*os.File{inr, outw, outw},
-	}
-
-	socketWriterError := make(chan error, 1)
-	defer close(socketWriterError)
+	lastTimestamp := int64(0)
+	previousEvents := strset.New()
+	writePending := true
```

Comment:
Do you mind walking me through the logic around setting/using `writePending` tomorrow?